File: solver/sage_replacement.py

import math
import random
from fractions import Fraction

# --- Helper Functions ---

def inverse_mod(a, m):
    """
    Compute the modular multiplicative inverse of a modulo m.
    Uses Python's pow(a, -1, m) which is efficient.
    """
    return pow(a, -1, m)
    
def getPrime(n):
    """
    Mimic Crypto.Util.number.getPrime if not available.
    """
    try:
        from Crypto.Util.number import getPrime as gp
        return gp(n)
    except ImportError:
        def is_prime(num):
             if num < 2: return False
             for i in range(2, int(num**0.5) + 1):
                 if num % i == 0: return False
             return True
        while True:
             num = random.getrandbits(n)
             if is_prime(num): return num

class GF:
    def __init__(self, p):
        self.p = p

# --- Matrix and LLL ---

class Matrix:
    def __init__(self, ring, rows, cols=None):
        if cols is None and isinstance(rows, list):
            self.mat = [r[:] for r in rows]
            self.rows = len(rows)
            self.cols = len(rows[0]) if self.rows > 0 else 0
        elif isinstance(rows, int) and isinstance(cols, int):
            self.rows = rows
            self.cols = cols
            self.mat = [[0]*cols for _ in range(rows)]
        else:
            raise ValueError("Invalid Matrix constructor arguments")
            
    def __getitem__(self, idx):
        if isinstance(idx, tuple):
            r, c = idx
            return self.mat[r][c]
        return self.mat[idx]

    def __setitem__(self, idx, val):
        if isinstance(idx, tuple):
            r, c = idx
            self.mat[r][c] = val
        else:
            self.mat[idx] = val

    def __repr__(self):
        return "Matrix([\n  " + ",\n  ".join(str(r) for r in self.mat) + "\n])"
    
    def __len__(self):
        return self.rows

    def LLL(self, delta=0.99):
        """
        Pure Python implementation of the Lenstra-Lenstra-Lovász (LLL) lattice reduction algorithm.
        Returns a NEW Matrix object with the reduced basis.
        Uses fractions.Fraction for exact Gram-Schmidt to handle large integers.
        """
        # Working with lists for speed
        B = [list(row) for row in self.mat]
        n = len(B)
        if n == 0: return Matrix(ZZ, [])
        k = 1
        
        def dot(v1, v2):
            return sum(x*y for x, y in zip(v1, v2))

        # Orthogonalized basis (Fraction)
        B_star = [[Fraction(x) for x in b] for b in B] 
        mu = [[Fraction(0)]*n for _ in range(n)]
        
        # Initial Gram-Schmidt
        for i in range(n):
            for j in range(i):
                d = dot(B_star[j], B_star[j])
                if d == 0: mu[i][j] = Fraction(0)
                else: mu[i][j] = dot(B[i], B_star[j]) / d
                for l in range(len(B[i])):
                    B_star[i][l] -= mu[i][j] * B_star[j][l]

        def update_gs_for_k(k):
            # Recompute B_star[k] and mu[k][0...k-1] using current B[k]
            B_star[k] = [Fraction(x) for x in B[k]]
            for j in range(k):
                d = dot(B_star[j], B_star[j])
                if d == 0: mu[k][j] = Fraction(0)
                else: mu[k][j] = dot(B[k], B_star[j]) / d
                for l in range(len(B[k])):
                    B_star[k][l] -= mu[k][j] * B_star[j][l]

        while k < n:
            # Size reduction
            for j in range(k - 1, -1, -1):
                if abs(mu[k][j]) > Fraction(1, 2):
                    q = round(float(mu[k][j]))
                    for l in range(len(B[k])):
                        B[k][l] -= q * B[j][l]
                    
                    # Update internal GS state for k
                    # Since B[k] changed, B_star[k] and mu[k][j] are invalid
                    update_gs_for_k(k)
                    
                    # NOTE: Updating B_star[k] does NOT affect previous B_star[0..k-1]
                    # Nor does it affect mu[k][0..j-1] because they depend on B[k] projection onto previous B_stars which didn't change?
                    # Actually, if we change B[k], we change its projection.
                    # Standard LLL size reduction step usually updates B[k] and then updates mu[k][j] = mu[k][j] - q.
                    # But full recompute is safer against bugs.

            # Lovasz condition check
            d_k = dot(B_star[k], B_star[k])
            d_k_1 = dot(B_star[k-1], B_star[k-1])
            f_delta = Fraction(int(delta*100), 100)
            
            if d_k >= (f_delta - mu[k][k-1]**2) * d_k_1:
                k += 1
            else:
                # Swap B[k] and B[k-1]
                B[k], B[k-1] = B[k-1], B[k]
                
                # We must update Gram-Schmidt not just for k-1 and k, but propagate changes to mu
                # The efficient way is updating mu, but let's do the "slow but sure" Re-GS
                # Recomputing GS for k-1, then k
                update_gs_for_k(k-1)
                update_gs_for_k(k)
                
                # CRITICAL MISSING STEP IN PREVIOUS VERSION:
                # We must update mu[i][k-1] and mu[i][k] for all i > k
                # because B_star[k-1] and B_star[k] have changed!
                for i in range(k + 1, n):
                    # Update mu[i][k-1]
                    d_prev = dot(B_star[k-1], B_star[k-1])
                    if d_prev == 0: mu[i][k-1] = Fraction(0)
                    else: mu[i][k-1] = dot(B[i], B_star[k-1]) / d_prev
                    
                    # Update mu[i][k]
                    d_curr = dot(B_star[k], B_star[k])
                    if d_curr == 0: mu[i][k] = Fraction(0)
                    else: mu[i][k] = dot(B[i], B_star[k]) / d_curr

                k = max(k - 1, 1)

        return Matrix(ZZ, B)

# Dummy object for ZZ ring
class RingZZ:
    pass
ZZ = RingZZ()

# --- Elliptic Curve ---

class EllipticCurve:
    def __init__(self, field, params):
        self.field = field
        self.a = params[0]
        self.b = params[1]
        self.p = field.p

    def __call__(self, x, y):
        return Point(self, x, y)

class Point:
    def __init__(self, curve, x, y):
        self.curve = curve
        self.x = x
        self.y = y
        self.p = curve.p
        if (x is not None) and (y is not None):
            # Check point on curve if desired
            pass

    def __eq__(self, other):
        if other is None: return False
        return self.x == other.x and self.y == other.y

    def __add__(self, other):
        if other is None: return self
        P = self
        Q = other
        p = self.p
        
        if P.x == Q.x and P.y != Q.y:
            return Point(self.curve, None, None) 

        if P.x == Q.x:
            if P.y == 0: return Point(self.curve, None, None)
            lam = (3 * P.x * P.x + self.curve.a) * inverse_mod(2 * P.y, p)
            lam %= p
        else:
            lam = (Q.y - P.y) * inverse_mod(Q.x - P.x, p)
            lam %= p

        x3 = (lam * lam - P.x - Q.x) % p
        y3 = (lam * (P.x - x3) - P.y) % p
        return Point(self.curve, x3, y3)

    def __rmul__(self, scalar):
        result = None
        addend = self
        scalar = int(scalar)
        while scalar > 0:
            if scalar & 1:
                if result is None: result = addend
                elif result.x is None: result = addend 
                elif addend.x is None: pass
                else: result = result + addend
            if addend.x is not None:
                addend = addend + addend
            scalar >>= 1
        return result

    def xy(self):
        return (self.x, self.y)

File: solver/sage_replacement.js

// BigInt Fraction for precise Gram-Schmidt
class BigFraction {
    constructor(n, d = 1n) {
        if (typeof n === 'number') n = BigInt(Math.round(n));
        if (typeof d === 'number') d = BigInt(Math.round(d));

        this.n = BigInt(n);
        this.d = BigInt(d);
        if (this.d < 0n) {
            this.n = -this.n;
            this.d = -this.d;
        }
    }

    add(other) {
        return new BigFraction(this.n * other.d + other.n * this.d, this.d * other.d);
    }

    sub(other) {
        return new BigFraction(this.n * other.d - other.n * this.d, this.d * other.d);
    }

    mul(other) {
        return new BigFraction(this.n * other.n, this.d * other.d);
    }

    div(other) {
        return new BigFraction(this.n * other.d, this.d * other.n);
    }

    absGt(val) {
        let an = this.n < 0n ? -this.n : this.n;
        return an * val.d > val.n * this.d;
    }

    round() {
        let n = this.n;
        let d = this.d;
        let abs_n = n < 0n ? -n : n;
        let rem = abs_n % d;
        let quot = abs_n / d;
        if (rem * 2n >= d) quot++;
        return n < 0n ? -quot : quot;
    }

    sq() {
        return new BigFraction(this.n * this.n, this.d * this.d);
    }
}

class Matrix {
    constructor(rows) {
        this.data = rows.map(r => r.map(c => BigInt(c)));
        this.rows = rows.length;
        this.cols = this.rows > 0 ? rows[0].length : 0;
    }

    dotFrac(v1, v2) {
        let sum = new BigFraction(0n);
        for (let i = 0; i < v1.length; i++) sum = sum.add(v1[i].mul(v2[i]));
        return sum;
    }

    lll(delta = 0.75) {
        let B = this.data.map(r => [...r]);
        const n = B.length;
        let k = 1;

        // B_star (Fraction vectors)
        let B_star = B.map(row => row.map(v => new BigFraction(v)));
        let mu = Array(n).fill(0).map(() => Array(n).fill(new BigFraction(0n)));

        const updateGS = (k) => {
            B_star[k] = B[k].map(v => new BigFraction(v));
            for (let j = 0; j < k; j++) {
                let d = this.dotFrac(B_star[j], B_star[j]);
                if (d.n === 0n) mu[k][j] = new BigFraction(0n);
                else {
                    let dotVal = new BigFraction(0n);
                    for (let l = 0; l < B[k].length; l++) {
                        dotVal = dotVal.add(B_star[j][l].mul(new BigFraction(B[k][l])));
                    }
                    mu[k][j] = dotVal.div(d);
                }
                for (let l = 0; l < this.cols; l++) {
                    B_star[k][l] = B_star[k][l].sub(mu[k][j].mul(B_star[j][l]));
                }
            }
        }

        for (let i = 0; i < n; i++) updateGS(i);

        const half = new BigFraction(1n, 2n);
        const f_delta = new BigFraction(BigInt(Math.floor(delta * 100)), 100n);

        while (k < n) {
            for (let j = k - 1; j >= 0; j--) {
                if (mu[k][j].absGt(half)) {
                    let q = mu[k][j].round();
                    for (let l = 0; l < this.cols; l++) B[k][l] -= q * B[j][l];
                    updateGS(k);
                }
            }

            let d_k = this.dotFrac(B_star[k], B_star[k]);
            let d_k1 = this.dotFrac(B_star[k - 1], B_star[k - 1]);

            let lhs = f_delta.sub(mu[k][k - 1].sq()).mul(d_k1);

            if (d_k.n * lhs.d >= lhs.n * d_k.d) {
                k++;
            } else {
                [B[k], B[k - 1]] = [B[k - 1], B[k]];
                updateGS(k - 1);
                updateGS(k);
                for (let i = k + 1; i < n; i++) {
                    let d_prev = this.dotFrac(B_star[k - 1], B_star[k - 1]);
                    if (d_prev.n === 0n) mu[i][k - 1] = new BigFraction(0n);
                    else {
                        let dotVal = new BigFraction(0n);
                        for (let l = 0; l < B[i].length; l++) dotVal = dotVal.add(B_star[k - 1][l].mul(new BigFraction(B[i][l])));
                        mu[i][k - 1] = dotVal.div(d_prev);
                    }
                    let d_curr = this.dotFrac(B_star[k], B_star[k]);
                    if (d_curr.n === 0n) mu[i][k] = new BigFraction(0n);
                    else {
                        let dotVal = new BigFraction(0n);
                        for (let l = 0; l < B[i].length; l++) dotVal = dotVal.add(B_star[k][l].mul(new BigFraction(B[i][l])));
                        mu[i][k] = dotVal.div(d_curr);
                    }
                }
                k = Math.max(k - 1, 1);
            }
        }
        return new Matrix(B);
    }
}

class EllipticCurve {
    constructor(p, a, b) {
        this.p = BigInt(p);
        this.a = BigInt(a);
        this.b = BigInt(b);
    }
    point(x, y) { return new Point(this, x, y); }
}

class Point {
    constructor(curve, x, y) {
        this.curve = curve;
        this.x = x === null ? null : BigInt(x);
        this.y = y === null ? null : BigInt(y);
    }
    isInf() { return this.x === null; }
    add(other) {
        if (this.isInf()) return other;
        if (other.isInf()) return this;
        const p = this.curve.p;
        if (this.x === other.x && this.y !== other.y) return new Point(this.curve, null, null);
        let lam;
        if (this.x === other.x) {
            if (this.y === 0n) return new Point(this.curve, null, null);
            let num = (3n * this.x * this.x + this.curve.a) % p;
            let den = (2n * this.y) % p;
            lam = (num * modInverse(den, p)) % p;
        } else {
            let num = (other.y - this.y) % p;
            let den = (other.x - this.x) % p;
            if (den < 0n) den += p;
            if (num < 0n) num += p;
            lam = (num * modInverse(den, p)) % p;
        }
        let x3 = (lam * lam - this.x - other.x) % p;
        if (x3 < 0n) x3 += p;
        let y3 = (lam * (this.x - x3) - this.y) % p;
        if (y3 < 0n) y3 += p;
        return new Point(this.curve, x3, y3);
    }
    mul(scalar) {
        let sc = BigInt(scalar);
        let res = new Point(this.curve, null, null);
        let add = this;
        while (sc > 0n) {
            if (sc & 1n) res = res.add(add);
            add = add.add(add);
            sc >>= 1n;
        }
        return res;
    }
}

function modInverse(a, m) {
    let [m0, x0, x1] = [m, 0n, 1n];
    if (m === 1n) return 0n;
    while (a > 1n) {
        let q = a / m;
        [m, a] = [a % m, m];
        [x0, x1] = [x1 - q * x0, x0];
    }
    if (x1 < 0n) x1 += m0;
    return x1;
}

module.exports = { Matrix, EllipticCurve, modInverse };
